ert/data/record/_record.py

```diff
@@ -0,0 +1,234 @@
+import pathlib
+from abc import ABC
+from collections import deque
+from enum import Enum
+from typing import Dict, List, Mapping, MutableMapping, Optional, Tuple, Union
+
+from beartype import beartype
+from beartype.roar import BeartypeException  # type: ignore
+
+import ert
+
+number = Union[int, float]
+numerical_record_data = Union[
+    List[number],
+    Dict[str, number],
+    Dict[int, number],
+]
+blob_record_data = bytes
+record_data = Union[numerical_record_data, blob_record_data]
+record_collection = Tuple["Record", ...]
+RecordIndex = Tuple[Union[int, str], ...]
+
+
+def _build_record_index(
+    data: numerical_record_data,
+) -> RecordIndex:
+    if isinstance(data, MutableMapping):
+        return tuple(data.keys())
+    else:
+        return tuple(range(len(data)))
+
+
+class RecordValidationError(Exception):
+    pass
+
+
+class RecordType(str, Enum):
+    LIST_FLOAT = "LIST_FLOAT"
+    MAPPING_INT_FLOAT = "MAPPING_INT_FLOAT"
+    MAPPING_STR_FLOAT = "MAPPING_STR_FLOAT"
+    BYTES = "BYTES"
+
+
+class Record(ABC):
+    @property
+    def data(self) -> record_data:
+        pass
+
+    @property
+    def record_type(self) -> RecordType:
+        pass
+
+    @staticmethod
+    def create(
+        spec_or_record: Union[
+            Dict[str, bytes], Dict[str, numerical_record_data], "Record"
+        ]
+    ) -> "Record":
+        if isinstance(spec_or_record, Record):
+            return spec_or_record
+        data = spec_or_record["data"]
+        if isinstance(data, bytes):
+            return BlobRecord(data)
+        else:
+            index = spec_or_record.get("index")
+            if not isinstance(index, int):
+                index = None
+            return NumericalRecord(data=data, index=index)
+
+
+class BlobRecord(Record):
+    def __init__(self, data: blob_record_data) -> None:
+        self._record_type = RecordType.BYTES
+
+        try:
+            self._data = self._validate_data(data)
```

Comment:
Any reason for choosing to set the data as the return value of `_validate_data` instead of splitting the responsibility of setting the members and validating them?

By my very brief look at beartype right now it does not seem like a validator needs to return anything to function as intended 🤔

Reply:
I'm all for a split here, but beartype do validate the return value as well.

Reply:
I'll have a better look at beartype then... So they are validating both the input to the function and the return type afterwards? Is this to allow for the validation functions to potentially also transform the data? 🤔

Reply:
Not really, I think we should split it.